File: web/src/components/services/vacancy.ts
import { API_BASE } from './cfg';
import { supabase } from './cfg';

export interface VacancyData {
  vacancyId: string;
  projectId: string;
  projectTitle?: string;
  projectShortDesc?: string;
  title: string;
  role: string;
  description: string;
  requiredTags: string[];
  isOpen?: boolean;
  publishedAt?: string;
  membersCount?: number;
  ratingCount?: number;
}

export interface VacancySearchParams {
  search?: string;
  searchField?: string;
  sortField?: string;
  sortAsc?: boolean;
  showBannedOnly?: boolean;
}

export const vacanciesApi = {
  // Получить все открытые заявки
  getAll: async (params?: VacancySearchParams): Promise<VacancyData[]> => {
    const { data: { session } } = await supabase.auth.getSession();

    const query = new URLSearchParams();
    if (params?.search)
      query.set('search', params.search);
    if (params?.searchField)
      query.set('searchField', params.searchField);
    if (params?.sortField)
      query.set('sortField', params.sortField);
    if (params?.sortAsc !== undefined)
      query.set('sortAsc', String(params.sortAsc));
    if (params?.showBannedOnly !== undefined)
      query.set('showBannedOnly', String(params.showBannedOnly));

    const response = await fetch(`${API_BASE}/api/vacancy?${query.toString()}`, {
      headers: {
        'Authorization': `Bearer ${session?.access_token}`,
        'Content-Type': 'application/json'
      },
    });
    if (!response.ok) throw new Error('Ошибка загрузки заявок');
      return await response.json();
  },

  // Получить заявки проекта
  getByProject: async (projectId: string): Promise<VacancyData[]> => {
    const { data: { session } } = await supabase.auth.getSession();
    
    const response = await fetch(`${API_BASE}/api/vacancy/project/${projectId}`, {
      headers: { 
        'Authorization': `Bearer ${session?.access_token}`,
        'Content-Type': 'application/json' 
      },
    });

    if (!response.ok) throw new Error('Ошибка загрузки заявок проекта');
    return await response.json();
  },

  // Создать заявку
  create: async (dto: {
    projectId: string;
    title: string;
    role: string;
    description: string;
    requiredTags: string[];
  }): Promise<VacancyData> => {
    const { data: { session } } = await supabase.auth.getSession();
    
    const response = await fetch(`${API_BASE}/api/vacancy`, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${session?.access_token}`,
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(dto)
    });

    if (!response.ok) {
      const error = await response.json().catch(() => ({}));
      throw new Error(error.message || 'Ошибка создания заявки');
    }
    return await response.json();
  },

  // Обновить заявку
  update: async (vacancyId: string, dto: Partial<{
    title: string;
    role: string;
    description: string;
    requiredTags: string[];
    isOpen: boolean;
  }>): Promise<VacancyData> => {
    const { data: { session } } = await supabase.auth.getSession();
    
    const response = await fetch(`${API_BASE}/api/vacancy/${vacancyId}`, {
      method: 'PUT',
      headers: {
        'Authorization': `Bearer ${session?.access_token}`,
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(dto)
    });

    if (!response.ok) {
      const error = await response.json().catch(() => ({}));
      throw new Error(error.message || 'Ошибка обновления заявки');
    }
    return await response.json();
  },

  // Удалить заявку
  delete: async (vacancyId: string): Promise<void> => {
    const { data: { session } } = await supabase.auth.getSession();
    
    const response = await fetch(`${API_BASE}/api/vacancy/${vacancyId}`, {
      method: 'DELETE',
      headers: {
        'Authorization': `Bearer ${session?.access_token}`,
        'Content-Type': 'application/json'
      },
    });

    if (!response.ok) throw new Error('Ошибка удаления заявки');
  },
};

File: api/Models/Vacancy.cs
using Supabase.Postgrest.Attributes;
using Supabase.Postgrest.Models;
using System;
using System.Collections.Generic;

namespace api.Models
{
    [Table("vacancies")]
    public class Vacancy : BaseModel
    {
        [PrimaryKey("vacancy_id", false)]
        public string VacancyId { get; set; } = string.Empty;

        [Column("project_id")]
        public string ProjectId { get; set; } = string.Empty;

        [Column("title")]
        public string Title { get; set; } = string.Empty;

        [Column("role")]
        public string Role { get; set; } = string.Empty;

        [Column("description")]
        public string Description { get; set; } = string.Empty;

        [Column("required_tags")]
        public List<string> RequiredTags { get; set; } = [];

        [Column("is_open")]
        public bool IsOpen { get; set; } = true;

        [Column("published_at")]
        public DateTime? PublishedAt { get; set; }
    }

    public class CreateVacancyDto
    {
        public string ProjectId { get; set; } = string.Empty;
        public string Title { get; set; } = string.Empty;
        public string Role { get; set; } = string.Empty;
        public string Description { get; set; } = string.Empty;
        public List<string> RequiredTags { get; set; } = [];
    }

    public class UpdateVacancyDto
    {
        public string? Title { get; set; }
        public string? Role { get; set; }
        public string? Description { get; set; }
        public List<string>? RequiredTags { get; set; }
        public bool? IsOpen { get; set; }
    }
}